File: examples/domain-availability/check-domains.mjs
import { readFileSync, mkdirSync, writeFileSync, existsSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

function parseDotEnvFile(dotEnvText) {
  const envs = {};
  for (const line of dotEnvText.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;

    const eq = trimmed.indexOf("=");
    if (eq <= 0) continue;

    const key = trimmed.slice(0, eq).trim().replace(/^\uFEFF/, "");
    let value = trimmed.slice(eq + 1).trim();

    // Strip inline comments for unquoted values: KEY=value # comment
    if (!(value.startsWith('"') || value.startsWith("'"))) {
      const hash = value.indexOf("#");
      if (hash >= 0) value = value.slice(0, hash).trim();
    }

    // Strip surrounding quotes
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }

    if (key) envs[key] = value;
  }
  return envs;
}

function findDotEnvPath(startDir) {
  let cur = startDir;
  while (true) {
    const candidate = join(cur, ".env");
    if (existsSync(candidate)) return candidate;

    const parent = dirname(cur);
    if (parent === cur) return null;
    cur = parent;
  }
}

function loadDotEnv({ override = false } = {}) {
  const startDir = process.cwd();
  const dotEnvPath = findDotEnvPath(startDir) || findDotEnvPath(dirname(fileURLToPath(import.meta.url)));
  if (!dotEnvPath) return null;

  const envs = parseDotEnvFile(readFileSync(dotEnvPath, "utf8"));
  for (const [k, v] of Object.entries(envs)) {
    if (override || process.env[k] === undefined || process.env[k] === "") {
      process.env[k] = v;
    }
  }
  return dotEnvPath;
}

// Load repo .env if present (does not override existing env vars)
loadDotEnv();

function getEnv(name, fallback = undefined) {
  const val = process.env[name];
  return val === undefined || val === "" ? fallback : val;
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function parseDomainsFile(filePath) {
  const raw = readFileSync(filePath, "utf8");
  return raw
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l && !l.startsWith("#"));
}

function csvEscape(value) {
  if (value === null || value === undefined) return "";
  const str = String(value);
  if (/[",\n\r]/.test(str)) return `"${str.replace(/"/g, '""')}"`;
  return str;
}

function normalizeExtract(extract) {
  if (!extract) return null;
  if (typeof extract === "string") {
    try {
      return JSON.parse(extract);
    } catch {
      return { status: "unknown", notes: extract };
    }
  }
  if (typeof extract === "object") return extract;
  return null;
}

function inferAvailabilityFromMarkdown(markdown) {
  if (!markdown) return { availability: "unknown", price: null, notes: "No markdown" };

  const text = markdown.toLowerCase();
  let availability = "unknown";

  // Heuristics (best-effort)
  if (/(\bavailable\b|\bfor sale\b|\bregister\b)/i.test(markdown)) availability = "available";
  if (/(\btaken\b|\bregistered\b|\bunavailable\b)/i.test(markdown)) availability = "taken";

  const priceMatch = markdown.match(/\$[\d,]+(?:\.\d{2})?/);
  const price = priceMatch ? priceMatch[0] : null;

  return { availability, price, notes: "Parsed from markdown" };
}

async function firecrawlScrape({ apiUrl, apiKey, scrapePath, url, timeoutMs, waitForMs }) {
  const body = {
    url,
    formats: ["extract", "markdown"],
    extract: {
      prompt:
        "You are checking domain availability on a domain search results page. Return ONLY a JSON object with keys: availability (one of available/taken/unknown), price (string or null), and notes (string).",
      schema: {
        type: "object",
        properties: {
          availability: {
            type: "string",
            enum: ["available", "taken", "unknown"],
          },
          price: { type: ["string", "null"] },
          notes: { type: "string" },
        },
        required: ["availability", "price", "notes"],
        additionalProperties: false,
      },
    },
    waitFor: waitForMs,
    timeout: timeoutMs,
  };

  const res = await fetch(`${apiUrl}${scrapePath}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const json = await res.json().catch(() => null);
  if (!res.ok) {
    const msg = json?.error || json?.message || `HTTP ${res.status}`;
    const err = new Error(msg);
    err.status = res.status;
    err.response = json;
    throw err;
  }

  if (json?.success === false) {
    const msg = json?.error || "Unknown Firecrawl error";
    throw new Error(msg);
  }

  return json;
}

async function withRetries(fn, { retries, baseDelayMs }) {
  let lastErr;
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (attempt < retries) {
        await sleep(baseDelayMs * Math.pow(2, attempt));
      }
    }
  }
  throw lastErr;
}

async function runPool(items, worker, concurrency) {
  const results = new Array(items.length);
  let idx = 0;

  async function runner() {
    while (true) {
      const i = idx++;
      if (i >= items.length) return;
      results[i] = await worker(items[i], i);
    }
  }

  const workers = Array.from({ length: Math.max(1, concurrency) }, () => runner());
  await Promise.all(workers);
  return results;
}

async function main() {
  const repoRoot = resolve(process.cwd());
  const inputFile = resolve(repoRoot, "examples", "domain-availability", "domains.txt");
  const outDir = resolve(repoRoot, "examples", "domain-availability", "out");
  mkdirSync(outDir, { recursive: true });

  const apiKey = getEnv("FIRECRAWL_API_KEY") ?? getEnv("TEST_API_KEY");
  if (!apiKey) {
    console.error("Missing FIRECRAWL_API_KEY (or TEST_API_KEY) environment variable");
    process.exit(1);
  }

  const defaultApiUrl = (() => {
    const port = getEnv("PORT", "3002");
    // HOST is often 0.0.0.0 for servers; use localhost for client calls.
    return `http://localhost:${port}`;
  })();

  const apiUrl = getEnv("FIRECRAWL_API_URL", defaultApiUrl).replace(/\/$/, "");
  const scrapePath = getEnv("FIRECRAWL_SCRAPE_PATH", "/v1/scrape");

  const concurrency = Number(getEnv("DOMAIN_CHECK_CONCURRENCY", "3"));
  const perRequestDelayMs = Number(getEnv("DOMAIN_CHECK_DELAY_MS", "250"));
  const waitForMs = Number(getEnv("DOMAIN_CHECK_WAITFOR_MS", "5000"));
  const timeoutMs = Number(getEnv("DOMAIN_CHECK_TIMEOUT_MS", "60000"));
  const retries = Number(getEnv("DOMAIN_CHECK_RETRIES", "2"));

  const domains = parseDomainsFile(inputFile);
  if (domains.length === 0) {
    console.log(`No domains found in ${inputFile}`);
    return;
  }

  console.log(`Checking ${domains.length} domains via Firecrawl...`);
  console.log(`API: ${apiUrl}${scrapePath}`);

  const startedAt = new Date().toISOString();

  const results = await runPool(
    domains,
    async (domain, i) => {
      const query = encodeURIComponent(domain);
      const searchUrl = `https://instantdomainsearch.com/?q=${query}`;

      const base = {
        domain,
        searchUrl,
        index: i + 1,
        timestamp: new Date().toISOString(),
      };

      try {
        const response = await withRetries(
          () =>
            firecrawlScrape({
              apiUrl,
              apiKey,
              scrapePath,
              url: searchUrl,
              timeoutMs,
              waitForMs,
            }),
          { retries, baseDelayMs: 500 }
        );

        const data = response?.data || {};
        const extracted = normalizeExtract(data.extract);

        let availability = extracted?.availability || extracted?.status || "unknown";
        if (!["available", "taken", "unknown"].includes(availability)) availability = "unknown";

        let price = extracted?.price ?? null;
        if (price !== null && typeof price !== "string") price = String(price);

        let notes = extracted?.notes ?? "";
        if (typeof notes !== "string") notes = JSON.stringify(notes);

        if (availability === "unknown" && data.markdown) {
          const fallback = inferAvailabilityFromMarkdown(data.markdown);
          availability = fallback.availability;
          price = price ?? fallback.price;
          if (!notes) notes = fallback.notes;
        }

        if (perRequestDelayMs > 0) await sleep(perRequestDelayMs);

        console.log(`[${i + 1}/${domains.length}] ${domain}: ${availability}${price ? ` (${price})` : ""}`);

        return {
          ...base,
          ok: true,
          availability,
          price,
          notes,
          hasExtract: Boolean(data.extract),
          hasMarkdown: Boolean(data.markdown),
        };
      } catch (err) {
        if (perRequestDelayMs > 0) await sleep(perRequestDelayMs);

        const message = err?.message || String(err);
        console.log(`[${i + 1}/${domains.length}] ${domain}: ERROR (${message})`);

        return {
          ...base,
          ok: false,
          error: message,
        };
      }
    },
    concurrency
  );

  const finishedAt = new Date().toISOString();

  const jsonOut = {
    startedAt,
    finishedAt,
    apiUrl,
    scrapePath,
    count: results.length,
    successCount: results.filter((r) => r.ok).length,
    failureCount: results.filter((r) => !r.ok).length,
    results,
  };

  const resultsJsonPath = resolve(outDir, "results.json");
  writeFileSync(resultsJsonPath, JSON.stringify(jsonOut, null, 2), "utf8");

  const csvHeader = [
    "domain",
    "availability",
    "price",
    "notes",
    "ok",
    "error",
    "searchUrl",
    "timestamp",
  ];

  const csvLines = [csvHeader.join(",")];
  for (const r of results) {
    csvLines.push(
      [
        r.domain,
        r.ok ? r.availability : "",
        r.ok ? r.price ?? "" : "",
        r.ok ? r.notes ?? "" : "",
        r.ok,
        r.ok ? "" : r.error ?? "",
        r.searchUrl,
        r.timestamp,
      ]
        .map(csvEscape)
        .join(",")
    );
  }

  const resultsCsvPath = resolve(outDir, "results.csv");
  writeFileSync(resultsCsvPath, csvLines.join("\n"), "utf8");

  console.log("\nDone.");
  console.log(`- JSON: ${resultsJsonPath}`);
  console.log(`- CSV:  ${resultsCsvPath}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/domain-availability/check-domains.py
#!/usr/bin/env python3
"""Bulk domain availability checker (via Firecrawl).

Reads domains from domains.txt, scrapes InstantDomainSearch via Firecrawl, and writes:
- out/results.json
- out/results.csv

Environment variables:
- FIRECRAWL_API_KEY (required)
- FIRECRAWL_API_URL (default: http://localhost:3002)
- FIRECRAWL_SCRAPE_PATH (default: /v1/scrape)

Tuning:
- DOMAIN_CHECK_CONCURRENCY (default: 3)
- DOMAIN_CHECK_DELAY_MS (default: 250)
- DOMAIN_CHECK_WAITFOR_MS (default: 5000)
- DOMAIN_CHECK_TIMEOUT_MS (default: 60000)
- DOMAIN_CHECK_RETRIES (default: 2)
"""

from __future__ import annotations

import csv
import json
import os
import re
import time
import urllib.parse
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def env(name: str, default: str | None = None) -> str | None:
    v = os.getenv(name)
    if v is None or v == "":
        return default
    return v


def _parse_dotenv(text: str) -> dict[str, str]:
    out: dict[str, str] = {}
    for line in text.splitlines():
        s = line.strip()
        if not s or s.startswith("#"):
            continue
        if "=" not in s:
            continue
        key, value = s.split("=", 1)
        key = key.strip().lstrip("\ufeff")
        value = value.strip()

        # Strip inline comments for unquoted values: KEY=value # comment
        if not (value.startswith('"') or value.startswith("'")):
            if "#" in value:
                value = value.split("#", 1)[0].strip()

        # Strip surrounding quotes
        if (value.startswith('"') and value.endswith('"')) or (
            value.startswith("'") and value.endswith("'")
        ):
            value = value[1:-1]

        if key:
            out[key] = value
    return out


def _find_dotenv(start: Path) -> Path | None:
    cur = start
    while True:
        candidate = cur / ".env"
        if candidate.exists():
            return candidate
        if cur.parent == cur:
            return None
        cur = cur.parent


def load_dotenv(*, override: bool = False) -> Path | None:
    """Load repo .env into os.environ (best-effort).

    - Does not override existing env vars by default.
    - Searches upward from CWD and from this script's directory.
    """

    here = Path(__file__).resolve().parent
    dot = _find_dotenv(Path.cwd()) or _find_dotenv(here)
    if not dot:
        return None

    vals = _parse_dotenv(dot.read_text(encoding="utf-8"))
    for k, v in vals.items():
        if override or os.getenv(k) in (None, ""):
            os.environ[k] = v
    return dot


def parse_domains_file(path: Path) -> list[str]:
    domains: list[str] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        s = line.strip()
        if not s or s.startswith("#"):
            continue
        domains.append(s)

    # De-dupe while preserving order
    seen: set[str] = set()
    out: list[str] = []
    for d in domains:
        k = d.lower()
        if k in seen:
            continue
        seen.add(k)
        out.append(d)
    return out


def safe_run_id_from_iso(iso: str) -> str:
    # Windows-safe folder name
    # 2026-01-08T19:15:26.666Z -> 20260108-191526
    iso = re.sub(r"\.\d+Z$", "Z", iso)
    iso = re.sub(r"\+\d\d:\d\d$", "Z", iso)
    iso = iso.replace("-", "").replace(":", "").replace("T", "-")
    iso = re.sub(r"Z$", "", iso)
    return iso[:15]


def load_running_runs_json(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    try:
        parsed = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(parsed, list):
            return [x for x in parsed if isinstance(x, dict)]
        if isinstance(parsed, dict) and isinstance(parsed.get("results"), list):
            # Old format: single run object
            return [parsed]
        return []
    except Exception:
        return []


def build_checked_domain_set_from_runs(runs: list[dict[str, Any]]) -> set[str]:
    checked: set[str] = set()
    for run in runs:
        results = run.get("results")
        if not isinstance(results, list):
            continue
        for r in results:
            if not isinstance(r, dict):
                continue
            domain = r.get("domain")
            if isinstance(domain, str) and domain.strip():
                checked.add(domain.strip().lower())
    return checked


def ensure_running_csv_has_v2_header(results_csv_path: Path, header_v2: list[str]) -> None:
    if not results_csv_path.exists():
        return

    try:
        first_line = results_csv_path.read_text(encoding="utf-8").splitlines()[0]
    except Exception:
        return

    if first_line.strip() == ",".join(header_v2):
        return

    legacy_header = ",".join(
        ["domain", "availability", "price", "notes", "ok", "error", "searchUrl", "timestamp"]
    )
    if first_line.strip() != legacy_header:
        return

    lines = [l for l in results_csv_path.read_text(encoding="utf-8").splitlines() if l]
    out_lines = [",".join(header_v2)]
    for line in lines[1:]:
        # Prefix empty run metadata for historical rows
        out_lines.append(",".join(["", "", "", line]))
    results_csv_path.write_text("\n".join(out_lines) + "\n", encoding="utf-8")


def append_rows_to_running_csv(results_csv_path: Path, header_v2: list[str], rows: list[list[Any]]) -> None:
    if not results_csv_path.exists():
        results_csv_path.write_text(",".join(header_v2) + "\n", encoding="utf-8")

    ensure_running_csv_has_v2_header(results_csv_path, header_v2)

    # Repair any previously glued rows caused by missing trailing newline.
    repair_running_csv_glued_rows(results_csv_path)

    first_line = results_csv_path.read_text(encoding="utf-8").splitlines()[0]
    if first_line.strip() != ",".join(header_v2):
        raise RuntimeError(f"Unexpected CSV header in {results_csv_path}; refusing to append.")

    if not rows:
        return

    with results_csv_path.open("a", encoding="utf-8", newline="") as f:
        w = csv.writer(f)
        for row in rows:
            w.writerow(row)


def repair_running_csv_glued_rows(results_csv_path: Path) -> None:
    if not results_csv_path.exists():
        return

    raw = results_csv_path.read_text(encoding="utf-8")
    if not raw:
        return

    lines = raw.splitlines()
    out: list[str] = []
    changed = False

    glued_marker = re.compile(r"(\d{8}-\d{6},20\d{2}-\d{2}-\d{2}T)")
    for line in lines:
        m = glued_marker.search(line)
        if m and m.start() > 0:
            changed = True
            out.append(line[: m.start()])
            out.append(line[m.start() :])
        else:
            out.append(line)

    if not changed:
        return

    text = "\n".join(out)
    if not text.endswith("\n"):
        text += "\n"
    results_csv_path.write_text(text, encoding="utf-8")


def normalize_extract(extract: Any) -> dict[str, Any] | None:
    if extract is None:
        return None
    if isinstance(extract, dict):
        return extract
    if isinstance(extract, str):
        try:
            parsed = json.loads(extract)
            return parsed if isinstance(parsed, dict) else {"availability": "unknown", "notes": extract}
        except Exception:
            return {"availability": "unknown", "notes": extract}
    return None


def infer_from_markdown(markdown: str | None, domain: str) -> dict[str, Any]:
    if not markdown:
        return {"availability": "unknown", "price": None, "notes": "No markdown"}

    lines = markdown.splitlines()
    domain_needle = f"[{domain}]("
    domain_line_index = next((i for i, l in enumerate(lines) if domain_needle in l or domain in l), -1)
    if domain_line_index < 0:
        return {"availability": "unknown", "price": None, "notes": "Domain not found in markdown"}

    window_start = max(0, domain_line_index - 10)
    window_end = min(len(lines), domain_line_index + 40)
    window_lines = lines[window_start:window_end]
    window_text = "\n".join(window_lines).lower()

    availability = "unknown"
    if "make offer" in window_text or "whois" in window_text:
        availability = "taken"
    if "continue" in window_text:
        availability = "available"

    import re

    price: str | None = None
    for i in range(domain_line_index, min(len(lines), domain_line_index + 12)):
        line = lines[i].strip()

        # If we hit another domain-like link before a price, stop.
        if i != domain_line_index and re.match(r"^\[[^\]]+\]\([^)]*\)", line) and "." in line and domain not in line:
            break

        m = re.search(r"\$[\d,]+(?:\.\d{2})?", line)
        if m:
            price = m.group(0)
            break

    notes_parts: list[str] = []
    if availability == "available":
        notes_parts.append("Found 'Continue' near domain")
    if availability == "taken":
        notes_parts.append("Found 'Make offer/WHOIS' near domain")
    notes_parts.append("Found price near domain" if price else "No domain-specific price found")

    return {"availability": availability, "price": price, "notes": "; ".join(notes_parts)}


@dataclass(frozen=True)
class Settings:
    api_key: str
    api_url: str
    scrape_path: str
    concurrency: int
    delay_ms: int
    waitfor_ms: int
    timeout_ms: int
    retries: int


def firecrawl_scrape(settings: Settings, url: str) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "url": url,
        "formats": ["extract", "markdown"],
        "extract": {
            "prompt": (
                "You are checking domain availability on a domain search results page. "
                "Return ONLY a JSON object with keys: availability (one of available/taken/unknown), "
                "price (string or null), and notes (string)."
            ),
            "schema": {
                "type": "object",
                "properties": {
                    "availability": {"type": "string", "enum": ["available", "taken", "unknown"]},
                    "price": {"type": ["string", "null"]},
                    "notes": {"type": "string"},
                },
                "required": ["availability", "price", "notes"],
                "additionalProperties": False,
            },
        },
        "waitFor": settings.waitfor_ms,
        "timeout": settings.timeout_ms,
    }

    endpoint = f"{settings.api_url}{settings.scrape_path}"
    req = urllib.request.Request(
        endpoint,
        data=json.dumps(payload).encode("utf-8"),
        method="POST",
        headers={
            "Authorization": f"Bearer {settings.api_key}",
            "Content-Type": "application/json",
        },
    )

    try:
        with urllib.request.urlopen(req, timeout=max(5, settings.timeout_ms // 1000 + 5)) as resp:
            raw = resp.read().decode("utf-8", errors="replace")
            data = json.loads(raw)
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8", errors="replace")
        try:
            data = json.loads(raw)
            msg = data.get("error") or data.get("message") or f"HTTP {e.code}"
        except Exception:
            msg = f"HTTP {e.code}"
        raise RuntimeError(msg) from e

    if isinstance(data, dict) and data.get("success") is False:
        raise RuntimeError(data.get("error") or "Unknown Firecrawl error")

    return data


def with_retries(fn, *, retries: int, base_delay_s: float = 0.5):
    last_exc: Exception | None = None
    for attempt in range(retries + 1):
        try:
            return fn()
        except Exception as e:
            last_exc = e
            if attempt < retries:
                time.sleep(base_delay_s * (2**attempt))
    assert last_exc is not None
    raise last_exc


def main() -> int:
    # Load repo .env if present (does not override existing env vars)
    load_dotenv(override=False)

    repo_root = Path.cwd()
    folder = repo_root / "examples" / "domain-availability"
    input_file = folder / "domains.txt"
    out_dir = folder / "out"
    out_dir.mkdir(parents=True, exist_ok=True)

    runs_dir = out_dir / "runs"
    runs_dir.mkdir(parents=True, exist_ok=True)

    api_key = env("FIRECRAWL_API_KEY") or env("TEST_API_KEY")
    if not api_key:
        print("Missing FIRECRAWL_API_KEY (or TEST_API_KEY) environment variable")
        return 1

    default_api_url = f"http://localhost:{env('PORT', '3002')}"
    api_url = (env("FIRECRAWL_API_URL", default_api_url) or "").rstrip("/")
    scrape_path = env("FIRECRAWL_SCRAPE_PATH", "/v1/scrape") or "/v1/scrape"

    settings = Settings(
        api_key=api_key,
        api_url=api_url,
        scrape_path=scrape_path,
        concurrency=max(1, int(env("DOMAIN_CHECK_CONCURRENCY", "3") or "3")),
        delay_ms=max(0, int(env("DOMAIN_CHECK_DELAY_MS", "250") or "250")),
        waitfor_ms=max(0, int(env("DOMAIN_CHECK_WAITFOR_MS", "5000") or "5000")),
        timeout_ms=max(1000, int(env("DOMAIN_CHECK_TIMEOUT_MS", "60000") or "60000")),
        retries=max(0, int(env("DOMAIN_CHECK_RETRIES", "2") or "2")),
    )

    if not input_file.exists():
        print(f"Input file not found: {input_file}")
        return 1

    domains = parse_domains_file(input_file)
    if not domains:
        print(f"No domains found in {input_file}")
        return 0

    running_json_path = out_dir / "results.json"
    running_csv_path = out_dir / "results.csv"

    prior_runs = load_running_runs_json(running_json_path)
    checked_domains = build_checked_domain_set_from_runs(prior_runs)
    domains_to_check = [d for d in domains if d.lower() not in checked_domains]
    skipped_count = len(domains) - len(domains_to_check)

    print(f"Checking {len(domains_to_check)} domains via Firecrawl...")
    if skipped_count > 0:
        print(f"Skipping {skipped_count} already-checked domains (from out/results.json)")
    print(f"API: {settings.api_url}{settings.scrape_path}")

    started_at = utc_now_iso()
    run_id = safe_run_id_from_iso(started_at)
    run_out_dir = runs_dir / run_id
    run_out_dir.mkdir(parents=True, exist_ok=True)

    def worker(domain: str, index: int) -> dict[str, Any]:
        search_url = f"https://instantdomainsearch.com/?q={urllib.parse.quote(domain)}"
        base: dict[str, Any] = {
            "domain": domain,
            "searchUrl": search_url,
            "index": index,
            "timestamp": utc_now_iso(),
        }

        try:
            response = with_retries(
                lambda: firecrawl_scrape(settings, search_url),
                retries=settings.retries,
                base_delay_s=0.5,
            )

            data = (response or {}).get("data") or {}
            extracted = normalize_extract(data.get("extract"))

            availability = (extracted or {}).get("availability") or (extracted or {}).get("status") or "unknown"
            if availability not in ("available", "taken", "unknown"):
                availability = "unknown"

            price = (extracted or {}).get("price", None)
            if price is not None and not isinstance(price, str):
                price = str(price)

            notes = (extracted or {}).get("notes", "")
            if notes is None:
                notes = ""
            if not isinstance(notes, str):
                notes = json.dumps(notes)

            if data.get("markdown"):
                fallback = infer_from_markdown(data.get("markdown"), domain)

                if availability == "unknown":
                    availability = fallback["availability"]

                if price is None:
                    price = fallback["price"]
                elif fallback["price"] is None:
                    price = None
                    notes = f"{notes}; Dropped unverified price" if notes else "Dropped unverified price"
                elif str(price) != str(fallback["price"]):
                    price = fallback["price"]
                    notes = (
                        f"{notes}; Replaced price with domain-local price" if notes else "Replaced price with domain-local price"
                    )

                if not notes:
                    notes = fallback["notes"]

            if settings.delay_ms:
                time.sleep(settings.delay_ms / 1000.0)

            print(f"[{index}/{len(domains_to_check)}] {domain}: {availability}{f' ({price})' if price else ''}")

            return {
                **base,
                "ok": True,
                "availability": availability,
                "price": price,
                "notes": notes,
                "hasExtract": bool(data.get("extract")),
                "hasMarkdown": bool(data.get("markdown")),
            }
        except Exception as e:
            if settings.delay_ms:
                time.sleep(settings.delay_ms / 1000.0)
            msg = str(e)
            print(f"[{index}/{len(domains_to_check)}] {domain}: ERROR ({msg})")
            return {**base, "ok": False, "error": msg}

    results: list[dict[str, Any]] = [None] * len(domains_to_check)  # type: ignore[assignment]

    with ThreadPoolExecutor(max_workers=settings.concurrency) as pool:
        futures = {
            pool.submit(worker, domain, i + 1): i
            for i, domain in enumerate(domains_to_check)
        }
        for fut in as_completed(futures):
            i = futures[fut]
            results[i] = fut.result()

    finished_at = utc_now_iso()

    out_json = {
        "runId": run_id,
        "startedAt": started_at,
        "finishedAt": finished_at,
        "apiUrl": settings.api_url,
        "scrapePath": settings.scrape_path,
        "count": len(results),
        "successCount": sum(1 for r in results if r.get("ok")),
        "failureCount": sum(1 for r in results if not r.get("ok")),
        "skippedCount": skipped_count,
        "results": results,
    }

    run_json_path = run_out_dir / "results.json"
    run_json_path.write_text(json.dumps(out_json, indent=2), encoding="utf-8")

    run_csv_path = run_out_dir / "results.csv"
    with run_csv_path.open("w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["domain", "availability", "price", "notes", "ok", "error", "searchUrl", "timestamp"])
        for r in results:
            ok = bool(r.get("ok"))
            w.writerow(
                [
                    r.get("domain", ""),
                    r.get("availability", "") if ok else "",
                    r.get("price", "") if ok else "",
                    r.get("notes", "") if ok else "",
                    ok,
                    "" if ok else r.get("error", ""),
                    r.get("searchUrl", ""),
                    r.get("timestamp", ""),
                ]
            )

    # Append to running outputs
    updated_runs = prior_runs + [out_json]
    running_json_path.write_text(json.dumps(updated_runs, indent=2), encoding="utf-8")

    csv_header_v2 = [
        "runId",
        "runStartedAt",
        "runFinishedAt",
        "domain",
        "availability",
        "price",
        "notes",
        "ok",
        "error",
        "searchUrl",
        "timestamp",
    ]

    csv_rows_v2: list[list[Any]] = []
    for r in results:
        ok = bool(r.get("ok"))
        csv_rows_v2.append(
            [
                run_id,
                started_at,
                finished_at,
                r.get("domain", ""),
                r.get("availability", "") if ok else "",
                r.get("price", "") if ok else "",
                r.get("notes", "") if ok else "",
                ok,
                "" if ok else r.get("error", ""),
                r.get("searchUrl", ""),
                r.get("timestamp", ""),
            ]
        )

    append_rows_to_running_csv(running_csv_path, csv_header_v2, csv_rows_v2)

    print("\nDone.")
    print(f"- Run JSON: {run_json_path}")
    print(f"- Run CSV:  {run_csv_path}")
    print(f"- Running JSON: {running_json_path}")
    print(f"- Running CSV:  {running_csv_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
